more/auction/auction.js

```diff
@@ -0,0 +1,198 @@
+/* global E makePromise */
+// Copyright (C) 2019 Agoric, under Apache License 2.0
+
+import harden from '@agoric/harden';
+import { mustBeSameStructure } from '../../util/sameStructure';
+import { makeNatExtentOps } from '../../core/config/extentOps/natExtentOps';
+
+// A Seller will provide a good to be auctioned and a possibly empty purse to
+// show the currency in which bids must be expressed. The Auctioneer will create
+// seats which can offer bids. The highest bid will get the Seller's good, and a
+// refund of the difference between their bid and the second highest. The lower
+// bids will get their money back.
+const auction = {
+  start: (terms, inviteMaker) => {
+    const {
+      agencyEscrowInstallationP,
+      currencyAssetDesc,
+      goodsAssetDesc,
+      timerP,
+      deadline,
+      minBidCount,
+      minPrice,
+    } = terms;
+
+    const escrowTerms = harden({ currencyAssetDesc, goodsAssetDesc });
+    const escrowedGoods = makePromise();
+    const sellerWinnings = makePromise();
+    const sellerRefund = makePromise();
+
+    // map from bidderSeats to corresponding agencySeats we can close with.
+    const agencySeatsP = new Map();
+    let bestBidderP;
+    let bidderSeatCount = 0;
+    let bidsReceived = 0;
+    let bestPrice = 0;
+    let secondPrice = 0;
+    const currencyAssayP = currencyAssetDesc.label.assay;
+    const auctionComplete = makePromise();
+
+    // If bidder is undefined, cancel all Bids, otherwise cancel all but bidder.
+    function cancelExcept(bidder) {
+      // Cancel entries unless the key matches the argument
+      for (const [bidderSeatKey] of agencySeatsP) {
+        if (bidder === undefined || bidderSeatKey !== bidder) {
+          E(bidderSeatKey).cancel();
+        }
+      }
+    }
+
+    // By analogy with 'strictly greater than': x includes y and is not equal
+    function strictlyIncludes(leftAssetDesc, rightAssetDesc) {
+      // TODO(hibbert) look up strategy from assay with extentOpsLib
+      const natExtentOps = makeNatExtentOps();
+      return (
+        natExtentOps.includes(leftAssetDesc, rightAssetDesc) &&
+        !natExtentOps.equals(leftAssetDesc, rightAssetDesc)
+      );
+    }
+
+    E(timerP)
+      .delayUntil(deadline)
+      .then(() => {
+        // hold auction unless too few Bids or minPrice not met
+        if (
+          bidsReceived < minBidCount ||
+          strictlyIncludes(minPrice, secondPrice)
+        ) {
+          cancelExcept(undefined);
+          sellerRefund.res(escrowedGoods.p);
+          auctionComplete.reject(
+            bidsReceived < minBidCount
+              ? 'too few bids'
+              : 'minimum price not met.',
+          );
+          return;
+        }
+        cancelExcept(bestBidderP);
+        const bestBidAgencySeatP = agencySeatsP.get(bestBidderP);
+        const paidAssetDescP = E(bestBidAgencySeatP).consummateDeal(
+          bestPrice,
```

Comment:
I see. But you already know that it offered the best price. The consumeDeal API is better without this parameter.